pdftext/pdf/pages.py

```diff
@@ -0,0 +1,243 @@
+from __future__ import annotations
+
+import math
+import statistics
+from typing import List
+
+import pypdfium2 as pdfium
+
+from pdftext.pdf.chars import get_chars
+from pdftext.pdf.utils import flatten
+from pdftext.schema import Blocks, Chars, Line, Lines, Pages, Span, Spans
+
+
+def get_spans(chars: Chars) -> Spans:
+    spans: Spans = []
+    span: Span = None
+
+    def span_break():
+        return spans.append({
+            "bbox": char["bbox"],
+            "text": char["char"],
+            "rotation": char["rotation"],
+            "font": char["font"],
+            "char_start_idx": char["char_idx"],
+            "char_end_idx": char["char_idx"],
+            "chars": [char]
+        })
+
+    for char in chars:
+        if spans:
+            span = spans[-1]
+
+        if not span:
+            span_break()
+            continue
+
+        # we break on any change in font info
+        if any(char['font'][k] != span['font'][k] for k in ['name', 'flags', 'size', 'weight']):
+            span_break()
+            continue
+
+        if char['rotation'] != span['rotation']:
+            span_break()
+            continue
+
+        # we also break on hyphenation
+        if span['text'].endswith("\x02"):
+            span_break()
+            continue
+
+        span['text'] += char['char']
+        span['char_end_idx'] = char['char_idx']
+        span['bbox'] = span['bbox'].merge(char['bbox'])
+        span['chars'].append(char)
+
+    return spans
+
+
+def get_lines(spans: Spans) -> Lines:
+    lines: Lines = []
+    line: Line = None
+
+    def line_break(): return lines.append({"spans": [span], "bbox": span["bbox"], "rotation": span["rotation"]})
```

Comment:
return not needed, probably should unroll the function to make it easier to read